src/caffeine/coarray_access_s.f90

```diff
@@ -1,24 +1,66 @@
 ! Copyright (c), The Regents of the University of California
 ! Terms of use are as specified in LICENSE.txt
 submodule(coarray_access_m) coarray_access_s
+  use caffeine_h_m, only: caf_put, caf_get, as_int
+  use coarray_queries_m, only: prif_image_index
+  use iso_c_binding, only: c_loc
+  use prif_queries_m, only: prif_base_pointer
 
   implicit none
 
 contains
 
   module procedure prif_put
+    integer(c_int) :: image
+    integer(c_intptr_t) :: remote_base, remote_ptr
+
+    call prif_image_index(coarray_handle, coindices, image_index=image)
+    call prif_base_pointer(coarray_handle, image, remote_base)
+    remote_ptr = &
+        remote_base &
+        + (as_int(first_element_addr) &
+        - as_int(coarray_handle%info%coarray_data))
+    call prif_put_raw( &
+        image_num = image, &
+        local_buffer = c_loc(value), &
+        remote_ptr = remote_ptr, &
+        size = size(value) * coarray_handle%info%element_length)
   end procedure
 
   module procedure prif_put_raw
+    call caf_put( &
+        image = image_num, &
+        dest = remote_ptr, &
+        src = local_buffer, &
+        size = size)
   end procedure
 
   module procedure prif_put_raw_strided
   end procedure
 
   module procedure prif_get
+    integer(c_int) :: image
+    integer(c_intptr_t) :: remote_base, remote_ptr
+
+    call prif_image_index(coarray_handle, coindices, image_index=image)
+    call prif_base_pointer(coarray_handle, image, remote_base)
+    remote_ptr = &
+        remote_base &
+        + (as_int(first_element_addr) &
+        - as_int(coarray_handle%info%coarray_data))
```

Comment:
As previously expressed, I'm concerned about the superfluous performance overhead incurred by the `as_int()` helper invoked in this critical-path. As currently written, `prif_get()` requires four cross-language calls (prif_base_pointer->caf_convert_base_addr, 2x as_int and prif_get_raw->caf_get). All such calls should be expected to incur a full function call overhead at runtime and inhibit optimization of surrounding code. Ideally we'd replace `as_int` with a pure Fortran helper that the Fortran compiler can analyze and optimize the calls down to simple arithmetic.

In the absence of that ideal solution, we can still halve the number of  function call overheads by eliminating the calls to `as_int`: adding a C-side wrapper around `caf_get` that accepts the three pointer-sized arguments and performs the arithmetic to compute `remote_ptr` before calling the raw `caf_get`.

Reply:
This is now entered as issue #67 which proposes to resolve these using the "ideal solution" mentioned above.